spec/PeriodicBackgroundSync-index.bs

```diff
@@ -133,43 +146,81 @@ The user agent should cap the duration and frequency of these events to limit re
+
+    The user agent MAY include factors such as user engagement with the origin to decide this time interval, allowing origins with high user engagement to update their web apps more often. The user agent SHOULD also ensure this time interval conforms to the caps asserted by [=minimum periodic sync interval for any origin=] and the [=minimum periodic sync interval across origins=].
+
+    The user agent MAY also decide to [=retry=] each failed [=periodicsync event=] , with the count of retries capped to [=maximum number of retries=].
+  <div algorithm>
+    Calculating the [=periodic sync registration/time to fire=], |timeToFire| for |registration| involves running these steps:
+    1. If [=periodic sync registration/count of retries=] for |registration| is 0, <dfn>calculate the time to fire of the first attempt of</dfn> the [=periodicsync event=]:
+        1. Let |now|, a timestmap, represent the current time.
+        1. Let |origin| be the [=origin=] associated with |registration|'s [=periodic sync registration/service worker registration=].
+        1. Let |penalty| be a user agent defined penalty to account for the level of user engagement with the origin.
+        1. Let |minimumIntervalForOrigin| be |penalty|*[=minimum periodic sync interval for any origin=].
+        1. Let |delayForOrigin| be the multiple of |minimumIntervalForOrigin| greater than or equal to [=periodic sync registration/minimum interval=].
+        1. Let |timeTillScheduledEventForOrigin|, a number, be the [=periodic sync scheduler/time till the next scheduled periodicsync event=] for |origin|, if any, null otherwise.
+        1. If |timeTillScheduledEventForOrigin| is not null and |timeTillScheduledEventForOrigin| - |delayForOrigin| is less than |minIntervalForOrigin|:
+              1. If |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin|.
+              1. Else if |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|.
+        1. Let |timeSinceLastPeriodicSync| be null if [=periodic sync scheduler/time the last periodicsync event  was fired=] is null, else |now| - [=periodic sync scheduler/time the last periodicsync event was fired=].
+        1. If |timeSinceLastPeriodicSync| is null, set |timeToFire| to |delayForOrigin| + |now|.
+        1. Else:
+            1. If |timeSinceLastPeriodicSync| is greater than equal to [=minimum periodic sync interval across origins=], set |timeToFire| to |delayForOrigin| + |now|.
+            1. Else, set |timeTillNextAllowedPeriodicSync| to [=minimum periodic sync interval across origins=] - |timeSinceLastPeriodicSync|. Set |timeToFire| to the maximum of |delayForOrigin| + |now|, and |timeTillNextAllowedPeriodicSync| + |now|.
+        1. Set |registration|'s [=periodic sync registration/time to fire=] to |timeToFire|.
```

Comment:
I like this, thanks! It's simpler, gets rid of the |penalty| variable. I might add a reference to the global cap though (minimum periodic sync interval across origins)

The last fired time for a reg isn't tracked because every time we fire, we immediately calculate time to fire again, so last fired time ~= now. Do you see value in adding it?

Reply:
See my proposal for the global scheduler, that's where I apply the global cap.

It seems weird that sync can't immediately adjust based on engagement in any browser, but I trust that you folks have thought it through.

Reply:
Also, the suggestion here doesn't take into account timeTillScheduledEventForOrigin, which I think we need to include in the algorithm. Calculating a frequency for the origin is the first step, the second is to slot it in with existing scheduled events for the origin, such that the frequency is still ensured. Is it ok to be handwavey here about that?

Reply:
I catered for that in my proposal for the global scheduler I think.

Reply:
Ack. updated.